Blog post Super Optimization

SocratesWong opened the pull request.

SocratesWong committed:
Merge pull request #1 from sampsyo/2020fa

Upstream pull

SocratesWong committed:
Blog post

sampsyo commented:
Thanks for the draft, @SocratesWong! I'll review it soon, but in the mean time, please remember to make the Zulip post to start discussion (cf syllabus).

sampsyo commented on content/blog/2020-11-30-super-optimization/index.md:
```diff
@@ -0,0 +1,49 @@
+title = "Super Optimization"
+[extra]
+bio = """
+  Socrates Wong is a secound year PhD student at Cornell who intreset lies in Computer Archicture.
```
"secound" and "intreset" (please run spell check)

sampsyo commented on content/blog/2020-11-30-super-optimization/index.md:
```diff
@@ -0,0 +1,49 @@
++++
+
+## Background
+Most complier optimization (peephole, dead code elimination, constant propagation, loop invariant code motion, common sub expression elimination) focus on transforming the program in a systematic manner, and the step taken by the optimizer are backed by formal rigorous proofs that are both sound and complete.  Superoptimizer is a new radical way of preforming optimization.  Instead of taking the  ...
```
"sub expression" -> subexpression or sub-expression (your choice)

sampsyo replied:
"formal rigorous proofs" is quite optimistic! Very few compilers are backed by formal proofs.

sampsyo commented on content/blog/2020-11-30-super-optimization/index.md:
```diff
@@ -0,0 +1,49 @@
+
+## Background
+Most complier optimization (peephole, dead code elimination, constant propagation, loop invariant code motion, common sub expression elimination) focus on transforming the program in a systematic manner, and the step taken by the optimizer are backed by formal rigorous proofs that are both sound and complete.  Superoptimizer is a new radical way of preforming optimization.  Instead of taking the  ...
+The paper we are discussing this week is ["Superoptimizer"][superopt] – A look at the Smallest Program” (Henry), this paper explains the purpose of superoptimizer, which converts an optimizing problem to a search space, how to reduce the search space and comparison with other common optimization techniques that are common at the publication of the paper. 
```
The author's name is [Alexia Massalin](https://en.wikipedia.org/wiki/Alexia_Massalin) (she changed her name since the paper was published).

sampsyo commented on content/blog/2020-11-30-super-optimization/index.md:
```diff
@@ -0,0 +1,49 @@
+The paper we are discussing this week is ["Superoptimizer"][superopt] – A look at the Smallest Program” (Henry), this paper explains the purpose of superoptimizer, which converts an optimizing problem to a search space, how to reduce the search space and comparison with other common optimization techniques that are common at the publication of the paper. 
+
+
+## Super Optimization
```
Please pick either "superoptimization" or "Super Optimization" and stick with it consistently throughout the blog post. (Probably the former, not capitalized, because that's what the paper uses.)

sampsyo commented on content/blog/2020-11-30-super-optimization/index.md:
```diff
@@ -0,0 +1,49 @@
+
+
+## Super Optimization
+Super Optimization is finding the shortest program by exhaustive search inside a finite search space of the instruction set architecture (ISA).  Compare to other optimization techniques that focus on transformation and match, super optimization takes it to a whole different level.  Most optimizations techniques focus on the potential performance benefit on utilizing a transformation and ensuring  ...
```
The sentence beginning "Compare to…" is a comma splice. Replace the comma with a colon, semicolon, or new sentence.

sampsyo replied:
"optimizations techniques" -> optimization techniques

sampsyo replied:
The phrase "sample spaces" is not common and may need more explanation.

sampsyo replied:
"ombinations", "eah", "shortestlength"

sampsyo commented on content/blog/2020-11-30-super-optimization/index.md:
```diff
@@ -0,0 +1,49 @@
+## Super Optimization
+Super Optimization is finding the shortest program by exhaustive search inside a finite search space of the instruction set architecture (ISA).  Compare to other optimization techniques that focus on transformation and match, super optimization takes it to a whole different level.  Most optimizations techniques focus on the potential performance benefit on utilizing a transformation and ensuring  ...
+
+Then super optimization then preform a test of the generated program determine if the generated program is equivalent with the orgnial program and if they are the same, then generated programwill be the optimal porgram for the ISA.  Although this soulds simple, verifing the generated program is a non trival task.  One of the sloution proposed in the paper was boolean test.  However as the complex ...
```
"preform", "programwill", "soulds", "sloution"… [stopping reporting spelling errors now]

sampsyo replied:
Probably use math mode for "2^31"

sampsyo replied:
What is a "minterm"? Probably needs explanation. Overall, I'm not sure what the sentence "However as the complexity of some instructions have expressio of the order of 2^31 minterms" means (centrally, I'm not sure what "complexity" means in the context means).

sampsyo replied:
"a alternative"

sampsyo replied:
Can you please explain what a "boolean test" is? I don't think it's evident from context.

sampsyo replied:
Why does (Massalin) appear in parentheses here? Is this a quote from the paper? If so, be sure to use quotation marks. And then you can just say "(from the paper)".

sampsyo commented on content/blog/2020-11-30-super-optimization/index.md:
```diff
@@ -0,0 +1,49 @@
+
+Then super optimization then preform a test of the generated program determine if the generated program is equivalent with the orgnial program and if they are the same, then generated programwill be the optimal porgram for the ISA.  Although this soulds simple, verifing the generated program is a non trival task.  One of the sloution proposed in the paper was boolean test.  However as the complex ...
+
+The paper also has proposed a probailistict execution test in which the superoptimizer picks a test vectr and see if the output results are the same between the test and generated program.  Although this test is not compprehansive, the author claims that there was never a case where probailistict execution test passes while boolean test fails.  The time it takes for exhaustive search can be futhe ...
```
It's surprising that random tests never seem to fail in practice. Maybe a little discussion about how that's surprising and why that's true is in order?

sampsyo replied:
This paragraph mentions pruning but not how it works. How is it possible to prune some sequences—i.e., how does the superoptimizer know that the longer sequence is equivalent to some shorter sequence?

sampsyo commented on content/blog/2020-11-30-super-optimization/index.md:
```diff
@@ -0,0 +1,49 @@
+
+## Results
+
+One of the most intresting part of this paper is the results.  Although compare to other papers where they have a long compression of performance, superoptimizer reported it's result by comparing the execution of the input porgram and assembly.  For example, given the following signum program: 
```
What is "compression of performance"?

sampsyo replied:
"it's" -> its

sampsyo commented on content/blog/2020-11-30-super-optimization/index.md:
```diff
@@ -0,0 +1,49 @@
+
+Getting signum program to be immplemented in less then 9 assembly statments is already hard.  But superoptimizer managed to convert the program into a series of adds, subtracts, negation without any branches.  Resulting in the following results:
+<img src="superout.png" style="width: 30%;">
+By removing branches and the number of instructions, this immplemation has provide huge benfits to morden out of order processors (remove the need of branch predictions) and in both number of execution cycles.    Other results are similar interesting and that demestrated clearly the different then what a regular complier would generate.  
```
"then" -> than

sampsyo replied:
Can you elaborate a little bit on why you're convinced these output programs are unattainable for a conventional compiler? It would be interesting to hear your thoughts on what the limits of the traditional approach are.

sampsyo commented on content/blog/2020-11-30-super-optimization/index.md:
```diff
@@ -0,0 +1,49 @@
+
+## Limitations, Merits and Shortcomings
+
+One of the biggest limitaiton of super optimization would have been complier time cost and the benfits taht can be obtained at run time.  Although super optimization can ensure optimality, it does not nessciarity ensure a preformance bouns over exisitng solutions. That being said, this paper has a great impact in both llvm and gcc communities.  Despite the super optimization is not being actively ...
```
"One of the biggest limitaiton" -> limitations

sampsyo replied:
Capitalize LLVM and GCC.

sampsyo replied:
"Despite the super optimization is not being actively used" -> Despite that

sampsyo replied:
What are you referring to when you say "their own superoptimization implementation"? Are you thinking of specific projects? Providing links, citations, or background would be useful.

sampsyo replied:
When you say "complier time cost and the benfits taht can be obtained at run time," I certainly agree with the first part but I'm not sure about the second part. Quite clearly, superoptimization is _very_ expensive at compile time. But if you _could_ do it, it seems like the performance should always be really good, right? Is there a reason you're worried about the performance results for the final generated code?

sampsyo commented on content/blog/2020-11-30-super-optimization/index.md:
```diff
@@ -0,0 +1,49 @@
+
+## Connections to Modern Works
+
+Super optimizations desipte having good results has not be very wide adopted in for compliation compliers framework since this paper has been writen.  This concept has been reintroduced as papers such as [Automatic generation of peephole superoptimizers (Bansal, Aiken 2006)] [autopep]  Although the usage was wastly different from the orginal form of super optimization, this cencept managed to end ...
```
"compliation compliers framework" (not sure what noun phrase this is supposed to be)

sampsyo replied:
This Markdown syntax doesn't look like it works. (Delete the space.) Also, please put titles of papers in quotation marks.

sampsyo commented on content/blog/2020-11-30-super-optimization/index.md:
```diff
@@ -0,0 +1,49 @@
+## Discussion Questions
+
+- The landscape in the discipline of search has evolved greatly since the publication of the paper in 1987,  especially in aspects such as algorithms, and computational capacity of modern hardware.  To what extent does this affects superoptimizier, and what changes would see if this paper is written now?
+- In the discipline of search in computer science, there exist a type of search called heuristics.  To what extent do you think it is acceptable to this topic?
```
You might be interested in STOKE: http://stoke.stanford.edu

sampsyo commented on content/blog/2020-11-30-super-optimization/index.md:
```diff
@@ -0,0 +1,49 @@
+
+- The landscape in the discipline of search has evolved greatly since the publication of the paper in 1987,  especially in aspects such as algorithms, and computational capacity of modern hardware.  To what extent does this affects superoptimizier, and what changes would see if this paper is written now?
+- In the discipline of search in computer science, there exist a type of search called heuristics.  To what extent do you think it is acceptable to this topic?
+- •	Although the two most common complier infrastructure (LLVM and GCC) have a superoptimizier implemented for their infrastructure, it is not currently used in regular complier sequence.  To what extent is it due to the performance at compilation time, the maturity of super optimizer, the lack of interest of finding the most optimal program, or some other reason that is blocking it from being ad ...
```
```suggestion
- Although the two most common complier infrastructure (LLVM and GCC) have a superoptimizier implemented for their infrastructure, it is not currently used in regular complier sequence.  To what extent is it due to the performance at compilation time, the maturity of super optimizer, the lack of interest of finding the most optimal program, or some other reason that is blocking it from being adopted as a standard compilation tool?
```

sampsyo replied:
Throughout, s/complier/compiler/g.

sampsyo reviewed (commented):
This is a solid high-level overview. In general, I think slightly more specificity with actual examples could be helpful, but no need to worry about that now. Here are some writing suggestions. (Please carefully check for spelling and typos!)

sampsyo reviewed (changes requested):
(Oops, forgot to mark the review as "request changes.")

sampsyo commented:
Yo @SocratesWong, are you planning on revising this post?

SocratesWong commented:
Ops. I didn’t see the github notifications.

sampsyo commented:
Just another reminder about this. Are you planning on doing the revisions at some point?

SocratesWong committed:
Update index.md

Should have fixed all the errors.  I appoligized for the execssive spelling errors, thought the spell check was working when it wasn't.

sampsyo reviewed (approved):
Looks good; thanks!
